eventseries/src/main/matcher/tfidf_matcher.py

```diff
@@ -0,0 +1,143 @@
+                #         print(f"######SERIES######{array2_strings[array2_index]}")
+
+                if (
+                        self.matches_df.loc[
+                            self.matches_df['event'] == array1_strings[array1_index], 'event_series'].values[
+                            0]) == \
+                        array2_strings[array2_index]:
+                    #         if(matched_events_dict[array1_strings[array1_index]] >)
+                    true_positives += 1
+                else:
+                    false_positives += 1
+                partially_matched_events.append(array1_strings[array1_index])
+
+            # Series not matched to any event
+            #     false_negatives = len(series_titles) - len(series_distinct)
+            false_negatives = len(self.event_titles) - (true_positives + false_positives)
+
+            # print("Results with threshold: ", threshold)
+            # print("true positives: ", true_positives)
+            # print("false positives: ", false_positives)
+            # print("false negatives: ", false_negatives)
+            precision = true_positives / (true_positives + false_positives)
+            # print("Precision: ", precision)
+            recall = true_positives / (true_positives + false_negatives)
+            # print("Recall: ", recall)
+            f1_score = 2 * (precision * recall) / (precision + recall)
+            if f1_score > self.best_f1_score:
+                self.best_threshold = threshold
+                self.best_f1_score = f1_score
+            # print("F1-Score: ", f1_score)
+            # print("Number of partial matches: ", len(partially_matched_events))
+            # print()
+        print("#####BEST THRESHOLD#####")
+        print(self.best_threshold)
+
+    def wikidata_match(self, existing_matches):
+        partially_matched_events = []
+        events_file = os.path.join(os.path.abspath("resources"), "events_without_matches.json")
+        series_file = os.path.join(os.path.abspath("resources"), "event_series.json")
+        with open(events_file) as file:
+            events = json.load(file)
+            event_titles = [item['title'] for item in events if 'title' in item]
+        event_titles = [event for event in event_titles if event not in existing_matches]
+        with open(series_file) as file:
+            series = json.load(file)
+            series_titles = [item["title"]["value"] for item in series["results"]["bindings"] if "title" in item]
+
+        vectorizer = TfidfVectorizer(stop_words=list(text.ENGLISH_STOP_WORDS))
+        array1_strings = event_titles
+        array2_strings = series_titles
+        tfidf_matrix = vectorizer.fit_transform(event_titles + series_titles)
+
+        # Calculate cosine similarity between array1 and array2
+        similarity_matrix = cosine_similarity(tfidf_matrix[:len(array1_strings)], tfidf_matrix[len(array1_strings):])
+
+        # Threshold for partial match
+        threshold = self.best_threshold
```

Comment:
How should a user of this method know that `matcher()` has to be called before in order to find the best threshold?